### PR TITLE
r/(linux|windows)_virtual_machine(_scale_set): support for Force Delete

### LINKED ISSUE #11089: r/(linux|windows)_virtual_machine: support for Force Delete
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The Preview feature `Force Delete` for a Linux & Windows Virtual Machine is coming out of Preview and should be supported as a feature-toggle for the new virtual machine resources.

### New or Affected Resource(s)

* `azurerm_linux_virtual_machine`
* `azurerm_windows_virtual_machine`

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR introduces support for the opt-in Force Delete behaviour for the following resources:

* `azurerm_linux_virtual_machine`
* `azurerm_linux_virtual_machine_scale_set`
* `azurerm_windows_virtual_machine`
* `azurerm_windows_virtual_machine_scale_set`

This functionality is an optional behaviour during the deletion of a Virtual Machine where the machine will be hard powered-off and all devices forcibly removed from the Virtual Machine - which can both improve the shut-down time and work around issues within the Azure API where resources can remain attached to a Virtual Machine during deletion.

Whilst generally speaking that'd make sense to be the default behaviour - unfortunately the API makes this an opt-in behaviour at this point in time and errors if this preview functionality isn't enabled, so this will have to be defaulted off for the moment.

Fixes #11089

### CODE REVIEW

**Review by mbfrahry — COMMENTED** [review]
Hey @tombuildsstuff! I tried running the tests but I'm not sure if these new attributes have been added properly.

```
=== CONT  TestAccWindowsVirtualMachine_otherForceDelete
testing.go:620: Step 1/2 error: Error running pre-apply refresh:
Error: Unsupported argument
on config892966522/terraform_plugin_test.tf line 5, in provider "azurerm":
5:       force_delete = true
An argument named "force_delete" is not expected here.
--- FAIL: TestAccWindowsVirtualMachine_otherForceDelete (6.86s)
```

```
=== CONT  TestAccLinuxVirtualMachine_otherForceDelete
testing.go:620: Step 1/2 error: Error running pre-apply refresh:
Error: Unsupported argument
on config259639429/terraform_plugin_test.tf line 5, in provider "azurerm":
5:       force_delete = true
An argument named "force_delete" is not expected here.
--- FAIL: TestAccLinuxVirtualMachine_otherForceDelete (6.67s)
```

**Comment by grayzu** [comment]
Also @tombuildsstuff, can you adjust this so that it will skip the Power off all together when Force delete is enabled? If force delete is enabled, the intent is to simply have the VM be deleted without waiting for it to shutdown or power off.

**Comment by madewithsmiles** [comment]
@tombuildsstuff thanks for the quick turnaround. Please hold off merging this PR if possible. 
I am part of the service team and I am raising a few issues from this PR internally. I/@grayzu will follow up here once things are clear. Thanks.

**Comment by grayzu** [comment]
@tombuildsstuff, I agree that having separate flags for shutdown and forceDelete makes sense. The ask here is to optimize the ForceDelete so that it aligns with the design intentions.

To explain a bit further, there are three phases that can be configured as part of a VM(SS) delete:
1.  Whether or not to wait for the VM(SS) to shut down as part of the power off.
2.  Whether or not to wait for the VM(SS) to be powered off.
3.  Whether or not to use forceDelete when deleting the VM(SS).

The flag that we have now allows users to skip the shutdown as part of the power off process (no 1 above).  With this flag set, the VM will still be powered off before it is deleted.

The new flag enables forceDelete to be turned on as part of the VM(SS) delete process (no 3 above).

The desire is to have steps 1 & 2 above skipped when forceDelete is turned on so the VM(SS) is not shutdown or powered off, it is just deleted.

So in the end the ask is to have a flag for Shutdown that would allow customers to control wether or not to skip shutdown as part of the power off process and a flag for forceDelete that would short cut the process and go straight to delete the VM(SS).

**Review by katbyte — APPROVED** [review]
LGTM 👍

**Comment by tombuildsstuff** [comment]
@grayzu 

> So in the end the ask is to have a flag for Shutdown that would allow customers to control wether or not to skip shutdown as part of the power off process and a flag for forceDelete that would short cut the process and go straight to delete the VM(SS).

That's what the latest commit(s) do, there's now 2 flags available:

```
provider "azurerm" {
  features {
    virtual_machine {
      force_delete = true
      shutdown_before_deletion = false
    }
  }
}
```

which will give you that behaviour (skipping shutdown & "force delete") - and allow the inverse (for a shutdown with force delete, to allow the reliability benefits of Force Delete, when that OS disk ultimately needs to be reused, for example via capturing an image of the virtual machine)

File: azurerm/internal/features/defaults.go [hunk 1]
```diff
@@ -19,8 +19,13 @@ func Default() UserFeatures {
 		VirtualMachine: VirtualMachineFeatures{
 			DeleteOSDiskOnDeletion: true,
 			GracefulShutdown:       false,
+			// TODO: this should likely be defaulted true in 3.0
```
Comment (grayzu) [inline azurerm/internal/features/defaults.go:22]: This should not ever default to true. This is a new feature that should be an opt-in for users.

**Review by grayzu — CHANGES_REQUESTED** [review]
Thanks. I have added a couple of comments regarding force delete to ensure that we align with the intended behavior of the new feature.

**Review by mbfrahry — APPROVED** [review]
LGTM!

**Comment by grayzu** [comment]
Thanks @katbyte! Your changes look good to me.

**Pull request merged**